File: app/src/main/java/com/example/moviesearcher/ui/grid/MainGridAdapter.java
package com.example.moviesearcher.ui.grid;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.example.moviesearcher.R;
import com.example.moviesearcher.databinding.ItemMovieBinding;
import com.example.moviesearcher.model.data.Movie;

import java.util.ArrayList;
import java.util.List;

class MainGridAdapter extends RecyclerView.Adapter<MainGridAdapter.ViewHolder> {

    private final List<Movie> movieList = new ArrayList<>();
    private MovieClickListener listener;

    MainGridAdapter(MovieClickListener listener) {
        this.listener = listener;
    }

    interface MovieClickListener {
        void onMovieClicked(View v);
    }

    void updateMovieList(List<Movie> movieList) {
        if (movieList.size() == 0)
            this.movieList.clear();
        if (this.movieList.containsAll(movieList))
            return;
        this.movieList.addAll(movieList);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemMovieBinding view = DataBindingUtil.inflate(inflater, R.layout.item_movie, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.itemView.setMovie(movieList.get(position));
    }

    @Override
    public int getItemCount() {
        return movieList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private ItemMovieBinding itemView;

        ViewHolder(@NonNull ItemMovieBinding itemView) {
            super(itemView.getRoot());
            this.itemView = itemView;

            itemView.getRoot().setOnClickListener(v -> listener.onMovieClicked(v));
        }
    }
}

File: app/src/main/java/com/example/moviesearcher/ui/grid/MainGridAdapter.kt
package com.example.moviesearcher.ui.grid

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.moviesearcher.R
import com.example.moviesearcher.databinding.ItemMovieBinding
import com.example.moviesearcher.model.data.Movie

class MainGridAdapter(private val listener: MovieClickListener): RecyclerView.Adapter<MainGridAdapter.ViewHolder>() {

    private val movieList: MutableList<Movie> = mutableListOf()

    interface MovieClickListener {
        fun onMovieClicked(v: View?)
    }

    fun updateMovieList(movieList: List<Movie>) {
        if (movieList.isNullOrEmpty())
            this.movieList.clear()
        if (this.movieList.containsAll(movieList))
            return
        this.movieList.addAll(movieList)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view: ItemMovieBinding = DataBindingUtil.inflate(inflater, R.layout.item_movie, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.onBind(movieList[position])
    }

    override fun getItemCount(): Int {
        return movieList.size
    }

    inner class ViewHolder(itemView: ItemMovieBinding): RecyclerView.ViewHolder(itemView.root) {
        private val view = itemView

        fun onBind(movie: Movie){
            view.movie = movie
            view.root.setOnClickListener { listener.onMovieClicked(it) }
        }
    }
}